### PR TITLE
Translate 'Models' to Kotlin

### PR DESCRIPTION
**Pull request by dpgraham**
Used the Android Studio code conversion tool and then made fixes to resolve broken compilation + broken unit tests.

### CODE REVIEW

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/AppiumParams.kt [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+    fun getUriParameterValue(name: String): String? {
+        return uriParams[name]
+    }
+
+    private fun setUriParameterValue(name: String, value: String?) {
+        uriParams[name] = value
+    }
+
+    companion object {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/AppiumParams.kt]: https://kotlinlang.org/docs/reference/java-to-kotlin-interop.html

Please add `JvmStatic` and `const` where necessary

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/MobileBackdoorMethod.kt [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    private var args: List<BackdoorMethodArg>? = null
+
+    val rawArgs: List<BackdoorMethodArg>?
+        get() = if (args == null) {
+            emptyList()
+        } else args
+
+    val argumentTypes: Array<Class<*>?>
+        get() {
+            val rawArgs = rawArgs
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/MobileBackdoorMethod.kt]: what is done in this line?

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/MobileBackdoorMethod.kt [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+            val types = arrayOfNulls<Class<*>>(rawArgs!!.size)
+            for (i in rawArgs.indices) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/MobileBackdoorMethod.kt]: forEachIndexed

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/MobileBackdoorMethod.kt [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+                types[i] = BackdoorUtils.parseType(rawArgs[i].type!!)
+            }
+            return types
+        }
+
+    val arguments: Array<Any?>
+        get() {
+            val rawArgs = rawArgs
+            val parsedArgs = arrayOfNulls<Any>(rawArgs!!.size)
+            for (i in rawArgs.indices) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/MobileBackdoorMethod.kt]: forEachIndexed

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/MotionEventParams.kt [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    private var button: Int? = null
+
+
+    val androidButtonState: Int
+        @Throws(InvalidArgumentException::class)
+        get() = getAndroidButtonState(this.getButton())
+
+    fun getButton(): Int? {
+        // Left button is the default
+        return if (button == null) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/MotionEventParams.kt]: how about `?:` ?

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/ScrollToPageParams.kt [hunk 6]
```diff
@@ -14,22 +14,30 @@
...
-import androidx.annotation.Nullable;
+import com.google.gson.annotations.SerializedName
 
-import io.appium.espressoserver.lib.helpers.w3c.models.Actions;
+class ScrollToPageParams : AppiumParams() {
+    var scrollTo: ScrollTo? = null
+    var scrollToPage: Int? = null
+        private set
+    var smoothScroll: Boolean? = null
+        get() = if (field == null) false else field
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/ScrollToPageParams.kt]: `?:`

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/Session.kt [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        /**
+         * Create a global session. Only one session can run per server instance so throw an exception
+         * if one already is in progress
+         *
+         * @return Serializable Session object
+         * @throws SessionNotCreatedException Thrown if a Session is already running
+         */
+        @Synchronized
+        fun createGlobalSession(desiredCapabilities: DesiredCapabilities): Session? {
+            if (globalSession != null) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/Session.kt]: let

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/SetClipboardParams.kt [hunk 8]
```diff
@@ -14,20 +14,20 @@
...
-import javax.annotation.Nullable;
+import androidx.annotation.NonNull
+import androidx.annotation.Nullable
 
-@SuppressWarnings("unused")
-public class AlertParams extends AppiumParams {
-    private String buttonLabel = null;
+class SetClipboardParams : AppiumParams() {
+    private val contentType: ClipboardDataType? = null
+    @get:Nullable
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/SetClipboardParams.kt]: I assume `String?` already shows it is nullable

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/SourceDocument.kt [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+
+    private var serializer: XmlSerializer? = null
+    private var tmpXmlName: String? = null
+
+    constructor() : this(null, null) {}
+
+    @Throws(IOException::class)
+    private fun setAttribute(attrName: ViewAttributesEnum, attrValue: Any?) {
+        // Do not write attributes, whose values equal to null
+        if (attrValue != null) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/SourceDocument.kt]: let

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/SourceDocument.kt [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+    }
+
+    @Throws(IOException::class)
+    private fun recordAdapterViewInfo(adapterView: AdapterView<*>) {
+        val adapter = adapterView.adapter ?: return
+        val adapterCount = adapter.count
+        val adapterData = ArrayList<String>()
+        var isAdapterTypeSet = false
+        for (i in 0 until adapterCount) {
+            val adapterItem = adapter.getItem(i) ?: continue
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/SourceDocument.kt]: I like this syntax

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/SourceDocument.kt [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+            setAttribute(ViewAttributesEnum.TEXT, viewText.rawText)
+            setAttribute(ViewAttributesEnum.HINT, viewText.isHint)
+        }
+        setAttribute(ViewAttributesEnum.RESOURCE_ID, viewElement.resourceId)
+        setAttribute(ViewAttributesEnum.VIEW_TAG, viewElement.viewTag)
+        if (view is AdapterView<*>) {
+            recordAdapterViewInfo((view as AdapterView<*>?)!!)
+        }
+
+        if (viewMap != null) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/SourceDocument.kt]: let

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/SourceDocument.kt [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+        }
+        if (lastError is OutOfMemoryError) {
+            throw lastError
+        }
+        throw AppiumException(lastError)
+    }
+
+    private fun performCleanup() {
+        if (tmpXmlName != null) {
+            getApplicationContext<Context>().deleteFile(tmpXmlName)
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/SourceDocument.kt]: let

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/SourceDocument.kt [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+            } catch (e: IOException) {
+                throw AppiumException(e)
+            } finally {
+                performCleanup()
+                RESOURCES_GUARD.release()
+            }
+
+            // Get a list of elements that are associated with that node
+            val views = ArrayList<View>()
+            if (list != null) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/SourceDocument.kt]: let

Comment (mykola-mokhnach) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/SourceDocument.kt]: actually, this assignment can be done within one single reduce, like
`views = list?.reduce(...)`

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/Strategy.kt [hunk 14]
```diff
@@ -43,28 +42,23 @@
...
-    public static Strategy fromString(final String text) throws InvalidStrategyException {
-        if (text != null) {
-            for (final Strategy s : Strategy.values()) {
-                if (text.equalsIgnoreCase(s.strategyName)) {
-                    return s;
+    companion object {
+
+        @Throws(InvalidStrategyException::class)
+        fun fromString(text: String?): Strategy {
+            if (text != null) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/Strategy.kt]: let

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/ToastLookupParams.kt [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+ * limitations under the License.
+ */
+
+package io.appium.espressoserver.lib.model
+
+class ToastLookupParams : AppiumParams() {
+    var text: String? = null
+    var isRegexp: Boolean? = null
+        get() {
+            if (field == null) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/ToastLookupParams.kt]: `?:`

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/TouchAction.kt [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+    // Number of ms to add or subtract to a timeout so that it isn't the exact number
+    private val TIMEOUT_BUFFER: Long = 10
+
+    var action: ActionType? = null
+    var options: TouchActionOptions? = null
+
+    private val origin: Origin
+        get() {
+            val origin = Origin()
+            if (options!!.elementId != null) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/TouchAction.kt]: why not to use `?. ` ?

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/TouchAction.kt [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+
+        return Arrays.asList(moveAction, downAction, waitAction)
+    }
+
+    // If an action list has fewer than three actions, pad them with 'pauses' of 0 duration
+    private fun padActionsList(actions: List<Action>): List<Action> {
+        // One Jsonwp Touch Action maps to three W3C Actions
+        val W3C_ACTIONS_PER_TOUCH_ACTION = 3
+
+        val paddedActions = ArrayList<Action>(3)
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/TouchAction.kt]: how about using map?

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/TouchAction.kt [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+    companion object {
+
+        @Throws(AppiumException::class)
+        fun toW3CInputSources(touchActionsLists: List<List<TouchAction>>): List<InputSource> {
+            var touchInputIndex = 0
+
+            val inputSources = ArrayList<InputSource>()
+            for (touchActions in touchActionsLists) {
+                val w3cActions = ArrayList<Action>()
+                for (touchAction in touchActions) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/TouchAction.kt]: map

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/UiautomatorParams.kt [hunk 19]
```diff
@@ -80,29 +58,24 @@ public Integer getIndex() {
...
 
-        public static List<String> getValidStrategyNames() {
-            List<String> validStrategies = new ArrayList<>();
-            for (Strategy strategy : Strategy.values()) {
-                validStrategies.add(strategy.getName());
-            }
-            return validStrategies;
+            val validStrategyNames: List<String>
+                get() {
+                    val validStrategies = ArrayList<String>()
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/UiautomatorParams.kt]: map

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/UiautomatorParams.kt [hunk 20]
```diff
@@ -161,24 +134,19 @@ public String getName() {
...
-        public static List<String> getValidActionNames() {
-            List<String> validActions = new ArrayList<>();
-            for (Action action : Action.values()) {
-                validActions.add(action.getName());
-            }
-            return validActions;
+            val validActionNames: List<String>
+                get() {
+                    val validActions = ArrayList<String>()
+                    for (action in Action.values()) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/UiautomatorParams.kt]: map

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/ViewElement.kt [hunk 21]
```diff
@@ -0,0 +1,205 @@
...
+            return if (start < 0) {
+                nameCopy
+            } else nameCopy.substring(0, start)
+        }
+
+    val index: Int
+        get() {
+            val parentView = view.parent
+            if (parentView is ViewGroup) {
+                for (index in 0 until parentView.childCount) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/ViewElement.kt]: how about using `find`?

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/TouchAction.kt (hunk 17 above)
Comment (dpgraham) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/TouchAction.kt]: `paddedActions` is data that expects three `Action` objects in order

**Comment by dpgraham** [comment]
I think I made a mistake trying to refactor too much at once... going to try to do a piece-by-piece approach.

**Pull request closed without merging**